### PR TITLE
Add support for dip.rotate_2d

### PR DESCRIPTION
**Pull request by meshtag**
This PR intends to add support for 2D rotation in DIP dialect via dip.rotate_2d. Following tasks are to be strictly completed before its merger.
WIP Tasks : 
- [x] - Finalise interface via #13 
- [x] - Add [operation description](https://github.com/meshtag/buddy-mlir/blob/bb20823ba1ec4ec670366ff05bbcf9e74ddd2057/include/Dialect/DIP/DIPOps.td#L79) and enhance documentation throughout

### CODE REVIEW

**Comment by zhanghb97** [comment]
Please rebase the PR and  resolve the conflicts.

File: lib/Conversion/LowerDIP/LowerDIPPass.cpp [hunk 1]
```diff
@@ -569,11 +570,454 @@ class DIPCorr2DLowering : public OpRewritePattern<dip::Corr2DOp> {
...
+        Value iotaValIndex = builder.create<arith::AddIOp>(loc, iv, indexStart);
+        Value iotaVal = indexToF32(builder, loc, iotaValIndex);
+
+        builder.create<memref::StoreOp>(loc, iotaVal, tempMem, ValueRange{iv});
+        builder.create<AffineYieldOp>(loc);
+      });
+
+  return builder.create<vector::LoadOp>(loc, vecType, tempMem, ValueRange{c0});
+
+  // Value c0F32 = indexToF32(builder, loc, c0);
```
Comment (zhanghb97) [inline lib/Conversion/LowerDIP/LowerDIPPass.cpp:684]: Should the following code be removed?

File: lib/Conversion/LowerDIP/LowerDIPPass.cpp [hunk 2]
```diff
@@ -569,11 +570,454 @@ class DIPCorr2DLowering : public OpRewritePattern<dip::Corr2DOp> {
...
+  return F32ToIndex(builder, loc, centerRound);
+}
+
+// Equivalent of std::iota.
+Value iotaVec(OpBuilder &builder, Location loc, MLIRContext *ctx,
+              Value indexStart, Value strideVal, VectorType vecType, Value c0,
+              int64_t stride) {
+  // ToDo : Try to get rid of memref load/store, find less expensive ways of
+  // implementing
+  //        this function.
```
Comment (zhanghb97) [inline lib/Conversion/LowerDIP/LowerDIPPass.cpp:667]: This looks like a weird indentation format.

**Review by zhanghb97 — COMMENTED** [review]
Don't forget to add the description and command-line example to `README.md`.

**Review by zhanghb97 — COMMENTED** [review]
Please modify a small format issue. Others LGTM.

**Review by zhanghb97 — APPROVED** [review]
Thanks! I think the next steps of this direction should be:
- Use the unified MemRef container and Img container.
- Add docs for the algorithm.
- Find a good unified place for the helper function.

**Comment by meshtag** [comment]
Thanks for the review!

> Use the unified MemRef container and Img container.

I have tried doing this, the last time I checked, the `Img` container was introducing some extra warnings. We can discuss the nature of these warnings and how/if to eliminate them before using the modified container in examples. 

> Add docs for the algorithm.

Yes, this is an important task and I'll start working on it during my free time. Where do you think these docs should go? Shall we create a markdown file similar to 2D convolution in `buddy-mlir/docs/` ?

> Find a good unified place for the helper function.

Do you think `buddy-mlir/include/Dialect/DIP/utility.h` is a good place for utility functions to live?

**Comment by zhanghb97** [comment]
>  the Img container was introducing some extra warnings.

Maybe we can use an issue to track this.

> Shall we create a markdown file similar to 2D convolution in buddy-mlir/docs/ ?

Yes!

> Do you think buddy-mlir/include/Dialect/DIP/utility.h is a good place for utility functions to live?

I think we can add a `buddy-mlir/include/Utils/` dir. And we can put the generic helper functions into `buddy-mlir/include/Utils/Utils.h` and domain-specific helper functions into `buddy-mlir/include/Utils/DIPUtils.h`. WDYT?

**Comment by meshtag** [comment]
> I think we can add a buddy-mlir/include/Utils/ dir. And we can put the generic helper functions into buddy-mlir/include/Utils/Utils.h and domain-specific helper functions into buddy-mlir/include/Utils/DIPUtils.h. WDYT?

Works for me.

**Pull request merged**